torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -2276,50 +2278,70 @@ Tensor eig_backward(const std::vector<torch::autograd::Variable> &grads, const T
+  // in torch.linalg.eigh, and we simply work with symmetric matrices.
+
+  // This check just can be triggered in the backwards of torch.symeig
   TORCH_CHECK(eigenvectors,
-           "symeig_backward: Setting eigenvectors to false in torch.symeig doesn't compute eigenvectors ",
-           "and hence we cannot compute backward. Please use torch.symeig(eigenvectors=True)");
+           "Setting eigenvectors=False in torch.symeig doesn't compute the eigenvectors, ",
+           "and hence we cannot compute the backward. Please use torch.symeig(eigenvectors=True)");
 
   auto glambda = grads[0];
   auto gv = grads[1];
 
   auto vh = v.conj().transpose(-2, -1);
 
+  if (glambda.defined()) {
+    // glambda is real even when the matrix is complex
+    glambda = glambda.to(self.dtype());
+  }
+
   Tensor result;
   if (gv.defined()) {
-      Tensor F = lambda.unsqueeze(-2) - lambda.unsqueeze(-1);
-      F.diagonal(/*offset=*/0, /*dim1=*/-2, /*dim2=*/-1).fill_(INFINITY);
-      F.pow_(-1);
-      result = at::matmul(v, at::matmul(F * at::matmul(vh, gv), vh));
-  } else {
-      result = at::zeros_like(self, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
-  }
+    // Computes v @ (diag(glambda) + F*(vh @ gv)) @ vh as per the source above
+    Tensor F = lambda.unsqueeze(-2) - lambda.unsqueeze(-1);
+    if (at::GradMode::is_enabled()) {
+      // Avoids differentiating through at infinity when doing gradgrad
+      // 1 could be any number, as we are going to overwrite the diagonal
+      F.diagonal(/*offset=*/0, /*dim1=*/-2, /*dim2=*/-1).fill_(1);
+    }
+    // We compute first the multiplication F*(vh @ gv) and then we zero-out the diagonal.
+    // Otherwise, if vh @ gh has NaNs on the diagonal, these would be propagated to the result
+    // as NaN * 0 = NaN.
+    result =  at::matmul(vh, gv) / F;
+    if (glambda.defined()) {
+      result.diagonal(/*offset=*/0, /*dim1=*/-2, /*dim2=*/-1).copy_(glambda);
+    }
+    else {
+      result.diagonal(/*offset=*/0, /*dim1=*/-2, /*dim2=*/-1).zero_();
+    }
 
-  if (glambda.defined()) {
-    glambda = glambda.to(self.dtype());
-    // computes v @ diag(glambda) @ vh
-    Tensor glambda_term = at::matmul(v * glambda.unsqueeze(-2), vh);
-    if (at::inplaceIsVmapCompatible(result, glambda_term)) {
-      result.add_(glambda_term);
+    result = at::matmul(v, at::matmul(result, vh));
+  }
+  else {
+    if (glambda.defined()) {
+      // If we just glambda is defined, one matmul suffices
+      result = at::matmul(v * glambda.unsqueeze(-2), vh);
     } else {
-      result = result + glambda_term;
+      // If neither is defined, there's nothing to do
+      return at::zeros_like(self, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
     }
   }
+  // The result here is alreaday symmetric (resp. Hermitian), but we project it onto the
+  // symmetric (resp. Hermitian) matrices to account for numerical errors
   return result.add(result.conj().transpose(-2, -1)).mul_(0.5);
```

Comment:
I disagree that the gradient is symmetric, primarily because the matrix `F` is NOT symmetric. But we assume that the elements of the input matrix, for real symmetric case, for example, are equivalent, i.e. `a_ij == a_ji`, so, even if the resulting gradient result is NOT symmetric, we can symmetrize it.